### PR TITLE
add checkpoints to db

### PR DESCRIPTION
**Pull request by wakamex**
create new checkpoints table, queried directly from hyperdrive

### CODE REVIEW

File: elfpy/data/db_schema.py [hunk 1]
```diff
@@ -47,11 +48,14 @@ class WalletInfo(Base):
     sharePrice: Mapped[Union[float, None]] = mapped_column(Numeric, default=None)
 
 
+@dataclass
```
Comment (dpaiton) [inline elfpy/data/db_schema.py:51]: The `Base` class already has a [mixin](https://realpython.com/inheritance-composition-python/#mixing-features-with-mixin-classes) that makes it behave like a dataclass:
```
class Base(MappedAsDataclass, DeclarativeBase):
    """Base class to subclass from to define the schema"""
```

As described in the [docs](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#:~:text=The%20MappedAsDataclass%20mixin%20may%20be%20applied%20either%20to%20the%20Declarative%20Base%20class%20or%20any%20superclass%2C%20as%20in%20the%20example%20below%3A)

Can you explain why you additionally need the `@dataclass` decorator here? I don't understand what it accomplishes.

File: elfpy/data/postgres.py [hunk 2]
```diff
@@ -179,6 +197,26 @@ def add_pool_infos(pool_infos: list[PoolInfo], session: Session) -> None:
...
+    session: Session
+        The initialized session object
+    """
+    for checkpoint_info in checkpoint_infos:
+        session.add(checkpoint_info)
+    try:
+        session.commit()
+    except sqlalchemy.exc.DataError as err:  # type: ignore
+        session.rollback()
+        print(f"{checkpoint_infos=}")
```
Comment (dpaiton) [inline elfpy/data/postgres.py:216]: no prints

File: elfpy/data/postgres.py [hunk 3]
```diff
@@ -527,34 +606,25 @@ def get_latest_block_number(session: Session) -> int:
...
+        The latest block number from the specified table
     """
-    query_results = session.query(Transaction).order_by(Transaction.id.desc()).first()
-    # If the table is empty, query_results will return None
-    if query_results is None:
+    result = session.execute(text(f'SELECT MAX("{table}"."blockNumber") FROM {table}')).fetchone()
+    if result is None:
         return 0
-    return int(query_results.blockNumber)
+    return int(result[0] or 0)
```
Comment (dpaiton) [inline elfpy/data/postgres.py:630]: why do you need `or 0` if you have the `if result is None` conditional?

File: elfpy/hyperdrive_interface/hyperdrive_interface.py [hunk 4]
```diff
@@ -119,6 +119,38 @@ def get_hyperdrive_pool_info(web3: Web3, hyperdrive_contract: Contract, block_nu
...
+
+    Arguments
+    ---------
+    web3: Web3
+        web3 provider object
+    hyperdrive_contract: Contract
+        The contract to query the pool info from
+    block_number: BlockNumber
+        The block number to query from the chain
+    Returns
```
Comment (dpaiton) [inline elfpy/hyperdrive_interface/hyperdrive_interface.py:135]: add a newline above Returns

File: elfpy/data/postgres.py [hunk 5]
```diff
@@ -475,9 +553,9 @@ def get_agents(session: Session, start_block: int | None = None, end_block: int
     query = session.query(WalletInfo.walletAddress)
     # Support for negative indices
     if (start_block is not None) and (start_block < 0):
-        start_block = _get_latest_block_number_wallet_info(session) + start_block + 1
+        start_block = get_latest_block_number_from_table(WalletInfo.__tablename__, session) + start_block + 1
     if (end_block is not None) and (end_block < 0):
-        end_block = _get_latest_block_number_wallet_info(session) + end_block + 1
+        end_block = get_latest_block_number_from_table(WalletInfo.__tablename__, session) + end_block + 1
```
Comment (dpaiton) [inline elfpy/data/postgres.py:558]: Why did you change these from private functions?

File: elfpy/data/db_schema.py (hunk 1 above)
Comment (wakamex) [reply elfpy/data/db_schema.py:51]: that's explained in the description of the PR that preceded this (#676)
I built on it for this PR (you can see the first commit is just that PR)
that PR is a required here so tests don't break, as explained in that PR

Comment (wakamex) [reply elfpy/data/db_schema.py:51]: actually I just reviewd that PR, and removed the dataclass decorator doesn't introduce any pyright issues. may have been caused by something else I was doing while building the tests. good catch!

File: elfpy/data/postgres.py (hunk 3 above)
Comment (wakamex) [reply elfpy/data/postgres.py:630]: that first step checks `result` which can itself be None, and can also contain `(None,)` so this is a second None check on that first element

File: elfpy/data/postgres.py [hunk 6]
```diff
@@ -179,6 +197,25 @@ def add_pool_infos(pool_infos: list[PoolInfo], session: Session) -> None:
...
+        A list of Checkpoint objects to insert into postgres
+    session: Session
+        The initialized session object
+    """
+    for checkpoint_info in checkpoint_infos:
+        session.add(checkpoint_info)
+    try:
+        session.commit()
+    except sqlalchemy.exc.DataError as err:  # type: ignore
+        session.rollback()
```
[the remark this answers is not in the document]
Comment (slundqui) [reply elfpy/data/postgres.py:215]: In general, I think it's always good practice to rollback if error (this block of try/catch would probably benefit from pulling out into a custom context). The operation defaults to an noop if nothing needs to be done.

File: elfpy/data/acquire_data.py [hunk 7]
```diff
@@ -177,11 +177,22 @@ def main(
         # Query and add block_pool_info
         pool_info_dict = hyperdrive_interface.get_hyperdrive_pool_info(web3, hyperdrive_contract, block_number)
         # Set defaults
-        for key in db_schema.PoolInfo.__annotations__.keys():
-            if key not in pool_info_dict.keys():
+        for key in db_schema.PoolInfo.__annotations__:
```
Comment (slundqui) [inline elfpy/data/acquire_data.py:180]: Can you add a TODO here to abstract this out? With the changes to hyperdrive_interface, it's on my list to pull the conversion between the interface to the db object into various functions

File: elfpy/data/acquire_data.py [hunk 8]
```diff
@@ -225,6 +236,7 @@ def main(
                         continue
                 if block_pool_info:
                     postgres.add_pool_infos([block_pool_info], session)
+
```
Comment (slundqui) [inline elfpy/data/acquire_data.py:239]: @dpaiton I love this new line, really breaks these code blocks into it's logical components and it makes it much easier to read

File: tests/data/test_checkpoint.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        # autoincrementing ids without init, whereas postgres does this with no issues
+        # Hence, we explicitly add id here
+        timestamp = datetime.now()
+        checkpoint = CheckpointInfo(blockNumber=1, timestamp=timestamp)
+        postgres.add_checkpoint_infos([checkpoint], session)
+        session.commit()
+
+        retrieved_checkpoint = session.query(CheckpointInfo).filter_by(blockNumber=1).first()
+        assert retrieved_checkpoint is not None
+        # event_value retreieved from postgres is in Decimal, cast to float
```
Comment (slundqui) [inline tests/data/test_checkpoint.py:44]: Remove dead comments, same for functions below

File: tests/data/test_checkpoint.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+class TestCheckpointInterface:
+    """Testing postgres interface for checkpoint table"""
+
+    def test_latest_block_number(self, session):
+        """Testing retrevial of checkpoint via interface"""
+        checkpoint_1 = CheckpointInfo(blockNumber=1, timestamp=datetime.now())
+        postgres.add_checkpoint_infos([checkpoint_1], session)
+        session.commit()
+
+        latest_block_number = postgres.get_latest_block_number_from_table(CheckpointInfo.__tablename__, session)
```
Comment (slundqui) [inline tests/data/test_checkpoint.py:84]: Don't forget to fix this call to `get_latest_block_number_from_table` after you rebase from main

**Review by slundqui — APPROVED** [review]

File: elfpy/data/postgres.py (hunk 5 above)
Comment (dpaiton) [reply elfpy/data/postgres.py]: Ok, that makes sense. I recommend that you keep it private until this issue actually comes up. In other words, go ahead and hand-roll a new access function for the next few new tables. Once we have a collection of them we can discuss generalizing.

File: elfpy/data/postgres.py [hunk 11]
```diff
@@ -58,6 +58,24 @@ def build_postgres_config() -> PostgresConfig:
     return PostgresConfig(**arg_dict)
 
 
+def query_tables(session):
+    """Return a list of tables in the database."""
+    stmt = text(
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/data/postgres.py:63]: I need to upgrade my GPT with this knowledge

File: elfpy/data/acquire_data.py (hunk 8 above)
Comment (dpaiton) [reply elfpy/data/acquire_data.py:239]: Sure thing! As a rule of thumb we should be purposeful about using them sparingly & preferentially using line-comments to separate blocks. However, that is a "rule of thumb" -- if anyone feels strongly about a newline then it's fine with me.

File: elfpy/data/db_schema.py [hunk 12]
```diff
@@ -82,12 +84,26 @@ class PoolConfig(Base):
...
+    shortBaseVolume: Mapped[Union[float, None]] = mapped_column(Numeric, default=None)
+
+
 class PoolInfo(Base):
     """
     Table/dataclass schema for pool info
     Mapped class that is a data class on the python side, and an declarative base on the sql side.
     """
 
+    # pylint: disable=too-many-instance-attributes
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/data/db_schema.py]: I tried to use them sparingly, only applied at the class level. `too-many-instance-attributes` is only applied twice for PoolInfo and PoolConfig`... I just removed it without complaints. but they both definitely have more than 7... how does it pass??
in our .pylintrc:
```
# Maximum number of attributes for a class (see R0902).
max-attributes=7
```

File: elfpy/data/postgres.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/data/postgres.py:64]: is this comment acceptable? `inspector = inspect(session.bind)  # nice gadget`

File: elfpy/data/postgres.py [hunk 13]
```diff
@@ -58,6 +58,24 @@ def build_postgres_config() -> PostgresConfig:
...
+        "SELECT tablename FROM pg_catalog.pg_tables "
+        "WHERE schemaname != 'pg_catalog' AND schemaname != 'information_schema'"
+    )
+    result = session.execute(stmt)  # Execute the statement
+    tables = result.fetchall()  # Fetch all the results
+    return tables
+
+
+def drop_table(session, table_name):
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/data/postgres.py:73]: added ORM version just for fun

File: elfpy/data/postgres.py [hunk 14]
```diff
@@ -179,6 +197,25 @@ def add_pool_infos(pool_infos: list[PoolInfo], session: Session) -> None:
         raise err
 
 
+def add_checkpoint_infos(checkpoint_infos: list[CheckpointInfo], session: Session) -> None:
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/data/postgres.py:201]: all of them? okay
